operators/src/opencl/cl_program.rs

```diff
@@ -0,0 +1,578 @@
+    ) -> Result<()>
+    where
+        T: Pixel + ocl::OclPrm,
+    {
+        let buffer = Buffer::<T>::builder()
+            .queue(self.kernel.default_queue().expect("checked").clone())
+            .len(raster.data_container.len())
+            .build()
+            .context(error::OCL)?;
+
+        self.kernel
+            .set_arg(format!("OUT{}", idx), &buffer)
+            .context(error::OCL)?;
+
+        self.output_buffers.push(f(buffer));
+
+        Ok(())
+    }
+
+    fn set_arguments(&mut self, params: &CLProgramParameters) -> Result<()> {
+        ensure!(
+            params.input_rasters.iter().all(Option::is_some),
+            error::CLProgramUnspecifiedRaster
+        );
+        self.output_buffers.clear();
+
+        for (idx, raster) in params.input_rasters.iter().enumerate() {
+            let raster = raster.expect("checked");
+            call_generic_raster2d!(raster, raster => {
+            let buffer = Buffer::builder()
+                .queue(self.kernel.default_queue().expect("checked").clone())
+                .flags(MemFlags::new().read_only())
+                .len(raster.data_container.len())
+                .copy_host_slice(&raster.data_container)
+                .build().context(error::OCL)?;
+
+                self.kernel.set_arg(format!("IN{}",idx), buffer).context(error::OCL)?;
+            });
+        }
+
+        for (idx, raster) in params.output_rasters.iter().enumerate() {
+            let raster = raster.as_ref().expect("checked");
+            match raster {
+                TypedRaster2D::U8(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::U8)?
+                }
+                TypedRaster2D::U16(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::U16)?
+                }
+                TypedRaster2D::U32(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::U32)?
+                }
+                TypedRaster2D::U64(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::U64)?
+                }
+                TypedRaster2D::I8(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::I8)?
+                }
+                TypedRaster2D::I16(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::I16)?
+                }
+                TypedRaster2D::I32(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::I32)?
+                }
+                TypedRaster2D::I64(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::I64)?
+                }
+                TypedRaster2D::F32(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::F32)?
+                }
+                TypedRaster2D::F64(raster) => {
+                    self.add_output_buffer(idx, raster, OutputBuffer::F64)?
+                }
+            }
```

Comment:
Use a macro for that?

Reply:
this macro would only be used a single time. Maybe we can create some more generic macro for such cases where two enums must comply.

Reply:
yes. no macro needed here.